pkg/querier/querier.go

```diff
@@ -0,0 +1,151 @@
+// SPDX-FileCopyrightText: Copyright 2024 The Minder Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Package querier provides tools to interact with the Minder database
+package querier
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/mindersec/minder/internal/db"
+	"github.com/mindersec/minder/pkg/config/server"
+	"github.com/mindersec/minder/pkg/engine/selectors"
+	"github.com/mindersec/minder/pkg/eventer"
+	"github.com/mindersec/minder/pkg/profiles"
+	"github.com/mindersec/minder/pkg/ruletypes"
+)
+
+var (
+	// ErrQuerierMissing is returned when the querier is not initialized
+	ErrQuerierMissing = fmt.Errorf("querier is missing, possibly due to closed or committed transaction")
+	// ErrProfileSvcMissing is returned when the profile service is not initialized
+	ErrProfileSvcMissing = fmt.Errorf("profile service is missing")
+	// ErrRuleSvcMissing is returned when the rule service is not initialized
+	ErrRuleSvcMissing = fmt.Errorf("rule service is missing")
+)
+
+// Store interface provides functions to execute db queries and transactions
+type Store interface {
+	ProjectHandlers
+	RuleTypeHandlers
+	ProfileHandlers
+	BundleHandlers
+	BeginTx() (Querier, error)
+}
+
+// Querier interface provides functions to interact with the Minder database using transactions
+// Calling Commit() or Cancel() is necessary after using the querier.
+// Note that they act as a destructor for the transaction
+// so any further calls using the querier will result in an error.
+type Querier interface {
+	ProjectHandlers
+	RuleTypeHandlers
+	ProfileHandlers
+	BundleHandlers
+	Commit() error
+	Cancel() error
+}
+
+// querierType represents the database querier
+type querierType struct {
+	store      db.Store
+	tx         *sql.Tx
+	querier    db.ExtendQuerier
+	ruleSvc    ruletypes.RuleTypeService
+	profileSvc profiles.ProfileService
+}
+
+// Closer is a function that closes the database connection
+type Closer func()
+
+// Ensure Type implements the Querier interface
+var _ Querier = (*querierType)(nil)
+
+// New creates a new instance of the querier
+func New(ctx context.Context, config *server.Config) (Store, Closer, error) {
+	// Initialize the database
+	store, dbCloser, err := initDatabase(ctx, config)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to setup database: %w", err)
+	}
+	// Get a watermill event handler
+	evt, err := eventer.New(ctx, nil, &config.Events)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to setup eventer: %w", err)
+	}
+	// Return the new Type
+	return &querierType{
```

Comment:
Oh, this is cute -- you're using querierType as both `Store` and `Querier`.  I _think_ that won't hurt you; anyone who goes to cast one to the other gets what's coming to them.